repo/content/index/id.go

```diff
@@ -0,0 +1,191 @@
+package index
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/kopia/kopia/repo/hashing"
+)
+
+// IDPrefix represents a content ID prefix (empty string or single character between 'g' and 'z').
+type IDPrefix string
+
+// ValidateSingle returns an error if a given prefix is invalid.
+func (p IDPrefix) ValidateSingle() error {
+	switch len(p) {
+	case 0:
+		return nil
+	case 1:
+		if p[0] >= 'g' && p[0] <= 'z' {
+			return nil
+		}
+	}
+
+	return errors.Errorf("invalid prefix, must be empty or a single letter between 'g' and 'z'")
+}
+
+// ID is an identifier of content in content-addressable storage.
+type ID struct {
+	data [hashing.MaxHashSize]byte
+
+	// those 2 could be packed into one byte, but that seems like overkill
+	prefix byte
+	idLen  byte
+}
+
+// MarshalJSON implements JSON serialization.
+func (i ID) MarshalJSON() ([]byte, error) {
+	s := i.String()
+
+	// nolint:wrapcheck
+	return json.Marshal(s)
+}
+
+// UnmarshalJSON implements JSON deserialization.
+func (i *ID) UnmarshalJSON(v []byte) error {
+	var s string
+
+	if err := json.Unmarshal(v, &s); err != nil {
+		return errors.Wrap(err, "unable to unmarshal ID")
+	}
+
+	tmp, err := ParseID(s)
+	if err != nil {
+		return errors.Wrap(err, "invalid ID")
+	}
+
+	*i = tmp
+
+	return nil
+}
+
+// Hash returns the hash part of content ID.
+func (i ID) Hash() []byte {
+	return i.data[0:i.idLen]
+}
+
+// EmptyID represents empty content ID.
+var EmptyID = ID{} // nolint:gochecknoglobals
+
+// prefixStrings contains precomputed single-character strings for all valid prefixes 'g'..'z'
+// nolint:gochecknoglobals
+var prefixStrings [256]IDPrefix
+
+func init() {
+	for i := 'g'; i <= 'z'; i++ {
+		prefixStrings[i] = IDPrefix([]byte{byte(i)})
+	}
+}
+
+func (i ID) less(other ID) bool {
+	if i.prefix != other.prefix {
+		if other.prefix == 0 {
+			// value is g..z, other is a..f, so i > other
+			return false
+		}
+
+		return i.prefix < other.prefix
+	}
+
+	return bytes.Compare(i.data[:i.idLen], other.data[:other.idLen]) < 0
+}
+
+// String returns a string representation of ID.
+func (i ID) String() string {
+	return string(i.Prefix()) + hex.EncodeToString(i.data[:i.idLen])
+}
+
+// Prefix returns a one-character prefix of a content ID or an empty string.
+func (i ID) Prefix() IDPrefix {
+	return prefixStrings[i.prefix]
+}
+
+// comparePrefix returns the value of strings.Compare(i.String(), p) in an optimized manner.
+func (i ID) comparePrefix(p IDPrefix) int {
+	switch len(p) {
+	case 0:
+		// empty ID == "", otherwise greater
+		if i == EmptyID {
+			return 0
+		}
+
+		return 1
```

Comment:
Does it make sense to handle `case 1:`?

Say, checking `p.ValidateSingle == nil`, and then computing a result based on `i.prefix`, `p[0]` and `len(i.data) > 0`?